### ISSUE TITLE
test_pb.NewCredTestCase is insanely long and causes some intermittent failures

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3210]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3210|
|Type|defect|
|Created|2008-04-26 08:39:12Z|
|Branch|https://github.com/twisted/twisted/tree/test-pb-newcred-3210|

It uses some callLater(1) to close connections. We can also see unclean warnings sometimes (eg http://buildbot.twistedmatrix.com/builders/osx10.5-py2.4-select/builds/789/steps/trial/logs/problems)

Both may be linked.

<details><summary>Searchable metadata</summary>

```
trac-id__3210 3210
type__defect defect
reporter__therve therve
priority__normal normal
milestone__ 
branch__branches_test_pb_newcred_3210 branches/test-pb-newcred-3210
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1209199152000000 1209199152000000
changetime__1211901815000000 1211901815000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [23434]) Branching to 'test-pb-newcred-3210'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve removed owner|
|-|-|

I think this is better. I won 3 seconds (from 3.1 to 0.1), and I didn't see any unclean errors. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @therve|
|-|-|

Not bad.  There seems to be a lot of duplication surrounding that `cleanup` function/pattern.  Think you could factor it so that logic is defined only once?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

I've done as suggested :). That should be better.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve|
|-|-|

`remote_doItLater` could use a docstring.  So could `getFactoryAndRootObject`.  Nice refactoring on `_disconnect`. :)  Merge when ready.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set status to `closed`|
|-|-|

(In [23719]) Merge test-pb-newcred-3210

Author: therve
Reviewer: exarkun
Fixes [#3210](https://github.com/twisted/twisted/issues/3210)

Fix test_pb.NewCredTestCase to remove useless callLater calls and to correctly
clean created connections.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
